pkg/kubectl/cmd/top.go

```diff
@@ -0,0 +1,204 @@
+	"github.com/spf13/cobra"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/kubectl"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/kubectl/metricsutil"
+)
+
+// TopOptions is the start of the data required to perform the operation. As new fields are added, add them here instead of
+// referencing the cmd.Flags()
+type TopOptions struct {
+}
+
+var (
+	topLong = dedent.Dedent(`
+		Display Resource (CPU/Memory/Storage) usage of nodes or pods.
+
+		The top command allows you to see the resource consumption of the nodes or pods.
+		It downloads the usage metrics of a given resource (node/pod) via the Resource Metrics API.`)
+
+	topExample = dedent.Dedent(`
+		  # Show metrics for all nodes
+		  kubectl top nodes
+
+		  # Show metrics for a given node
+		  kubectl top node NODE_NAME
+
+		  # Show metrics for all pods in the given namespace
+		  kubectl top pods --namespace=NAMESPACE
+
+		  # Show metrics for a given pod and its containers
+		  kubectl top pod POD_NAME --containers
+
+		  # Show metrics for the pods defined by label name=myLabel
+		  kubectl top pods -l name=myLabel`)
+)
+
+type MetricsCmdParams struct {
+	namespace       string
+	name            string
+	allNamespaces   bool
+	printContainers bool
+	params          map[string]string
+	client          *metricsutil.HeapsterMetricsClient
+	printer         *metricsutil.TopCmdPrinter
+}
+
+var HandledResources []unversioned.GroupKind = []unversioned.GroupKind{
+	api.Kind("Node"),
+	api.Kind("Pod"),
+}
+
+func GetResourcesHandledByTop() []string {
+	keys := make([]string, 0)
+	for _, k := range HandledResources {
+		resource := strings.ToLower(k.Kind)
+		keys = append(keys, resource)
+	}
+	return keys
+}
+
+func NewCmdTop(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+	options := &TopOptions{}
+	validArgs := GetResourcesHandledByTop()
+	argAliases := kubectl.ResourceAliases(validArgs)
+
+	cmd := &cobra.Command{
+		Use:     "top TYPE [NAME | -l label] [flags]",
+		Short:   "Display Resource (CPU/Memory/Storage) usage of nodes or pods",
+		Long:    topLong,
+		Example: topExample,
+		Run: func(cmd *cobra.Command, args []string) {
+			err := RunTop(f, out, cmd, args, options)
+			cmdutil.CheckErr(err)
+		},
+		ValidArgs:  validArgs,
+		ArgAliases: argAliases,
+	}
+	cmd.Flags().StringP("selector", "l", "", "Selector (label query) to filter on")
+	cmd.Flags().Bool("containers", false, "If present, print usage of containers within a pod. Ignored if the requested resource type is \"node\".")
+	cmd.Flags().Bool("all-namespaces", false, "If present, list the requested object(s) across all namespaces. Namespace in current context is ignored even if specified with --namespace.")
+	cmdutil.AddInclude3rdPartyFlags(cmd)
+	return cmd
+}
+
+func RunTop(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string, options *TopOptions) error {
+	if len(args) == 0 {
```

Comment:
Not a strong requirement, but I'd suggest to follow the `Complete`, `Validate`, `Run` pattern documented [here](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/kubectl-conventions.md#command-implementation-conventions) and move these validations. Also, we usually validate and error out flags that only apply to some scenarios, something like `--containers can't be used when the resource type is 'node'`.